Week2/homework/index.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+{
+  const readline = require('readline-sync');
+  const program = require('commander');
+  const { promisify } = require('util');
+  const mysql = require('mysql');
+
+  const connection = mysql.createConnection({
+    host: 'localhost',
+    user: 'hyfuser',
+    password: 'hyfpassword',
+    database: 'new_world',
+  });
+
+  const executeQuery = promisify(connection.query.bind(connection));
+
+  function exitProgram() {
+    connection.end();
+    process.exit();
+  }
+
+  async function makeQuery(query, input1, input2) {
+    try {
+      const result = await executeQuery(query, [input1, input2]);
+
+      if (result.length !== 0) {
+        console.table(result);
+      } else {
+        console.error(`The value you entered is invalid, please enter valid value!`);
+      }
+      exitProgram();
+    } catch (error) {
+      console.error(`There was an error: ${error.code} - ${error.message}`);
+    }
+  }
+
+  async function findCapitalCity() {
+    const country = readline.question('Enter a country: ');
+    const query = `select ci.name as capital from city ci inner join country co on ci.id = co.capital where co.name = ? `;
+    await makeQuery(query, country);
+  }
+
+  async function listLanguages() {
+    const region = readline.question('Enter a region: ');
+    const query = `select distinct cl.language from countrylanguage cl inner join country c on cl.countryCode = c.code where c.region = ?`;
+    console.log(`The languages spoken in ${region}: `);
+    await makeQuery(query, region);
+  }
+
+  async function findNumberOfCities() {
+    try {
+      const language = readline.question('Enter a language:');
+      const result = await executeQuery(
+        `select count(distinct ci.name) as numberOfCities from city ci 
+          join countrylanguage cl on ci.countryCode = cl.countryCode 
+          where cl.language = ?`,
+        language,
+      );
+
+      const [{ numberOfCities }] = result;
+
+      if (numberOfCities === 0) {
+        console.log(`Couldn't find ${language}, please enter valid language!`);
+      } else {
+        console.table(result);
+      }
+      exitProgram();
+    } catch (error) {
+      console.error(
+        `There was error while calculating number of cities: ${error.code} - ${error.message}`,
+      );
+    }
+  }
+
+  async function findCountries() {
+    try {
+      const region = readline.question('Enter a region: ');
+      const language = readline.question('Enter a language: ');
+      const countries = await executeQuery(
+        `select distinct c.name as countries from countrylanguage l 
+          inner join (select * from country where region = ? ) as c 
+          on c.code = l.countryCode 
+          and l.language = ? 
+          and l.isOfficial = 't'`,
```

Comment:
Yes it's good but I would have done :
select distinct c.name as countries from countrylanguage l 
          inner join country as c 
          on c.code = l.countryCode
          where
          c.region = ? 
          and l.language = ? 
          and l.isOfficial = 't'` ;

We usually join on the table and use it in the where clause

Reply:
I see your point, thanks for the explanation I will change mine as you suggested :+1: